src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java

```diff
@@ -548,6 +561,39 @@ public ProgrammingExercise updateProgrammingExercise(ProgrammingExercise program
         return savedProgrammingExercise;
     }
 
+    /**
+     * This method updates the build plan for the given programming exercise.
+     * It deletes the old build plan and creates a new one if the build plan configuration has changed.
+     *
+     * @param programmingExerciseBeforeUpdate the original programming exercise with its old values
+     * @param updatedProgrammingExercise      the changed programming exercise with its new values
+     */
+    private void updateBuildPlanForExercise(ProgrammingExercise programmingExerciseBeforeUpdate, ProgrammingExercise updatedProgrammingExercise) {
+        if (continuousIntegrationService.isEmpty()
+                || Objects.equals(programmingExerciseBeforeUpdate.getBuildPlanConfiguration(), updatedProgrammingExercise.getBuildPlanConfiguration())) {
+            return;
+        }
+        // we only update the build plan configuration if it has changed and is not null, otherwise we
+        // do not have a valid exercise anymore
+        if (updatedProgrammingExercise.getBuildPlanConfiguration() != null) {
+            if (!profileService.isLocalCi()) {
+                continuousIntegrationService.get().deleteProject(updatedProgrammingExercise.getProjectKey());
+                continuousIntegrationService.get().createProjectForExercise(updatedProgrammingExercise);
+                continuousIntegrationService.get().recreateBuildPlansForExercise(updatedProgrammingExercise);
+                resetAllStudentBuildPlanIdsForExercise(updatedProgrammingExercise);
+            }
+            if (buildScriptGenerationService.isPresent()) {
+                String script = buildScriptGenerationService.get().getScript(updatedProgrammingExercise);
+                updatedProgrammingExercise.setBuildScript(script);
+                programmingExerciseRepository.save(updatedProgrammingExercise);
+            }
```

Comment:
When Aeolus (`buildScriptGenerationService` ?) is used together with Jenkins, should the build script be created first, and then the build plan? Does it work this way round? I.e. is the correct script used to create the build plan?

Reply:
It works this way around as the build script itself isn't really used with jenkins enabled, but we generate the bash script so that we could switch to localci anytime.